perf/bps_tree.cc

```diff
@@ -0,0 +1,464 @@
+#include <stdint.h>
+#include <stdio.h>
+#include <stdbool.h>
+#include <inttypes.h>
+#include <time.h>
+#include <random>
+
+#include <benchmark/benchmark.h>
+#include <trivia/util.h>
+
+/* A simple test tree. */
+
+#define tree_i64_EXTENT_SIZE 2048
+#define tree_i64_elem_t int64_t
+#define tree_i64_key_t int64_t
+
+#define BPS_TREE_NAME tree_i64_t
+#define BPS_TREE_BLOCK_SIZE 512
+#define BPS_TREE_EXTENT_SIZE tree_i64_EXTENT_SIZE
+#define BPS_TREE_IS_IDENTICAL(a, b) ((a) == (b))
+#define BPS_TREE_COMPARE(a, b, arg) ((a) - (b))
+#define BPS_TREE_COMPARE_KEY(a, b, arg) ((a) - (b))
+#define bps_tree_elem_t tree_i64_elem_t
+#define bps_tree_key_t tree_i64_key_t
+#include "salad/bps_tree.h"
+#undef BPS_TREE_NAME
+#undef BPS_TREE_BLOCK_SIZE
+#undef BPS_TREE_EXTENT_SIZE
+#undef BPS_TREE_IS_IDENTICAL
+#undef BPS_TREE_COMPARE
+#undef BPS_TREE_COMPARE_KEY
+#undef bps_tree_elem_t
+#undef bps_tree_key_t
+
+/**
+ * Generate the benchmark variations required.
+ */
+
+/* Run the func benchmark on the type of the given size. */
+#define generate_benchmark_size(type, func, size) static void \
+	type##_##func##_size_##size(benchmark::State &state) \
+	{ \
+		test_##func<type>(state, size); \
+	} \
+	BENCHMARK(type##_##func##_size_##size)
+
+/* Run the func benchmark on the tree of the given height (max size). */
+#define generate_benchmark_height(type, func, height) static void \
```

Comment:
This will fail CI: "type" is a magic name for the checkpatch, for any another name it will complain about using macro arguments without parentheses.

The benchmark function name is passed partially for the same reason: if the name is used with ##, checkpatch will not require the parens.

This does not look right, but it's better than adding more exceptions to the checkpatch or making it parse C++ appropriately to determine if an argument is a template parameter or a function name that does not require wrapping by parens.

Foundamentally I would use something like clang-format for C style checks.

Reply:
> This will fail CI: "type" is a magic name for the checkpatch, for any another name it will complain about using macro arguments without parentheses.

Sorry, I didn't know about this nuance. Then just change `type` to `tree` in the comment to the macro above (`generate_benchmark_size`) please.

Reply:
An overlooked misspelling, thanks.